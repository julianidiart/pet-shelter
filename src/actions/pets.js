import uuid from "uuid";
import database from "../firebase";

// ADD_PET
export const addPet = pet => ({
  type: "ADD_PET",
  pet
});

export const startAddPet = (petData = {}) => {
  return dispatch => {
    const { name = "" } = petData;
    const pet = { name };
    database
      .ref("pet")
      .push(pet)
      .then(ref => {
        dispatch(addPet({ id: ref.key, ...pet }));
      });
  };
};

// REMOVE_PET
export const removePet = id => ({
  type: "REMOVE_PET",
  id
});

// EDIT_PET
export const editPet = (id, pet) => ({
  type: "EDIT_PET",
  id,
  pet
});
