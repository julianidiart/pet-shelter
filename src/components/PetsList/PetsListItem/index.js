import React from "react";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faMars,
  faVenus,
  faCheckCircle
} from "@fortawesome/free-solid-svg-icons";

export default ({ id, name, place, sex, breed, sterilized, size, color }) => {
  const gender =
    sex.toLowerCase() === "m" ? (
      <FontAwesomeIcon icon={faMars} title="male" />
    ) : (
      <FontAwesomeIcon icon={faVenus} title="female" />
    );
  return (
    <Link className="list-item" to={`/pets/${id}`}>
      <div>
        <h3 className="list-item__title">
          {name + " "} {gender}{" "}
          {sterilized && (
            <FontAwesomeIcon
              icon={faCheckCircle}
              title="sterilized"
              color="green"
            />
          )}
        </h3>
        <span className="list-item__sub-title">
          {breed + " " + (color ? "(" + color + ")" : "")}
        </span>
      </div>
      <div>
        <h3 className="list-item__place">{place}</h3>
      </div>
    </Link>
  );
};
