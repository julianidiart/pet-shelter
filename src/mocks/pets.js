export default [
  { id: "1", name: "Marla" },
  { id: "2", name: "Bimba" },
  { id: "3", name: "Aimee" },
  { id: "4", name: "Maia" },
  { id: "5", name: "Emma" },
  { id: "6", name: "Tommy" },
  { id: "7", name: "Bello Bruno" },
  { id: "8", name: "Bubu" },
  { id: "9", name: "Blaky" }
];
