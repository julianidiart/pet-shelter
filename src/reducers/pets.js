const petsReducerDefaultState = [];

export default (state = petsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_PET":
      return [...state, action.pet];
    case "REMOVE_PET":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_PET":
      return state.map(pet => {
        if (pet.id === action.id) {
          return {
            ...pet,
            ...action.pet
          };
        } else {
          return pet;
        }
      });
    case "SET_PETS":
      return action.pets;
    default:
      return state;
  }
};
