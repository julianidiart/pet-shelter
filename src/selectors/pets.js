import moment from "moment";

export default (pets, { text, sortBy, startDate, endDate, filter }) => {
  return pets
    .filter(pet => {
      const birthdateMoment = moment(pet.birthdate);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(birthdateMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(birthdateMoment, "day")
        : true;
      const textMatch =
        pet.name.toLowerCase().includes(text.toLowerCase()) ||
        pet.place.toLowerCase().includes(text.toLowerCase()) ||
        pet.breed.toLowerCase().includes(text.toLowerCase()) ||
        (pet.comments &&
          pet.comments.toLowerCase().includes(text.toLowerCase())) ||
        (pet.color
          ? pet.color.toLowerCase().includes(text.toLowerCase())
          : false) ||
        pet.chip
          .toString()
          .toLowerCase()
          .includes(text.toLowerCase());
      let filterMatch = true;
      const adoptedOrPassedAway = pet.adopted || pet.passedAway;
      switch (filter) {
        case "sterilized":
          filterMatch = !adoptedOrPassedAway && pet.sterilized === true;
          break;
        case "non-sterilized":
          filterMatch = !adoptedOrPassedAway && pet.sterilized === false;
          break;
        case "chipped":
          filterMatch = !adoptedOrPassedAway && pet.chip !== "";
          break;
        case "non-chipped":
          filterMatch = !adoptedOrPassedAway && pet.chip === "";
          break;
        case "adopted":
          filterMatch = pet.adopted === true;
          break;
        case "passed-away":
          filterMatch = pet.passedAway === true;
          break;
        default:
          filterMatch = !adoptedOrPassedAway;
      }

      return startDateMatch && endDateMatch && textMatch && filterMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.birthdate < b.birthdate ? 1 : -1;
      }
      return a.name < b.name ? 1 : -1;
    });
};
