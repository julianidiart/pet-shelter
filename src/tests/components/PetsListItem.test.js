import React from "react";
import { shallow } from "enzyme";
import PetsListItem from "../../components/PetsList/PetsListItem";
import pets from "../fixtures/pets";

test("should render PetsListItem correctly", () => {
  const wrapper = shallow(<PetsListItem {...pets[0]} />);
  expect(wrapper).toMatchSnapshot();
});
