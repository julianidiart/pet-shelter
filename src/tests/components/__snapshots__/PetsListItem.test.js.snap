// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PetsListItem correctly 1`] = `
<Link
  className="list-item"
  to="/pets/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Marla 
       
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              288,
              512,
              Array [],
              "f221",
              "M288 176c0-79.5-64.5-144-144-144S0 96.5 0 176c0 68.5 47.9 125.9 112 140.4V368H76c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h36v36c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12v-36h36c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-36v-51.6c64.1-14.5 112-71.9 112-140.4zm-224 0c0-44.1 35.9-80 80-80s80 35.9 80 80-35.9 80-80 80-80-35.9-80-80z",
            ],
            "iconName": "venus",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title="female"
        transform={null}
      />
    </h3>
    <span
      className="list-item__sub-title"
    >
      Incrocio (Black and white)
    </span>
  </div>
  <div>
    <h3
      className="list-item__place"
    >
      Barchetta
    </h3>
  </div>
</Link>
`;
